### ISSUE TITLE
[JS][Template] [TIME doesn't work properly with Template or AEL]

### ISSUE THREAD

**Issue by jwoo-msft** [microsoft/AdaptiveCards#3973]
# Platform

* JavaScript

# Version of SDK
latest from ts/renderer-refactor

# Details
Template used:  [Template.Functions.DateFunctions.json](https://github.com/microsoft/AdaptiveCards/blob/jwoo/Template-visualizer-dotnet/samples/Templates/Elements/Template.Functions.DateFunctions.json)
Template Data used: [Template.data.json](https://github.com/microsoft/AdaptiveCards/blob/jwoo/Template-visualizer-dotnet/samples/Templates/Elements/Template.data.json)
Expected: 5AM
Actual:12:00PM
![image](https://user-images.githubusercontent.com/4112696/80748794-5b732d80-8ada-11ea-8969-0ffa6b2a93ce.png)

**Comment by dclaux** [comment]
Fixed here: https://github.com/microsoft/AdaptiveCards/pull/3508/commits/a618dbfa819df0a546acd062ef8f5c87bc8c2772

**Comment by jwoo-msft** [comment]
It's still reproducible http://adaptivecards-ci.azurewebsites.net/designer/

{{TIME(${formatDateTime(date, 'yyyy-MM-ddTHH:mm:ssZ')})}} produces 12:00PM

{"date" : "03/15/2018 12:00:00"}

{{TIME(2018-03-15T12:00:00Z)}} produces 5AM

**Comment by dclaux** [comment]
Here is the deal:
- Our TIME function converts the date/time passed into the system's configured time zone
- Because of that, `{{TIME(2018-03-15T12:00:00Z)}}` produces 5AM PST which is correct
- The string "03/15/2018 12:00:00" doesn't contain any time zone information. So formatDataTime assumes it is expressed in the system's configured time zone, e.g. PST. And thus `${formatDateTime(date, 'yyyy-MM-ddTHH:mm:ssZ')}` produces 2018-03-15T12:00:00-07:00
- 2018-03-15T12:00:00-07:00 is obviously not the same date/time as 2018-03-15T12:00:00Z. The former is 12PM PST while the latter is 12PM UTC
- In turn, `{{TIME(2018-03-15T12:00:00-07:00)}}` produces 12PM as it should

Bottom line:
- There is no bug
- Never use date/time strings without time zone offset

**Comment by jwoo-msft** [comment]
@dclaux, thanks for the explanation

I ran a quick test.

"text": "{{TIME(${formatDateTime(date, 'yyyy-MM-ddTHH:mm:ss-07:00')})}}" produces 12:00 PM in the designer.
"text": "{{TIME(${formatDateTime(date, 'yyyy-MM-ddTHH:mm:ss-05:00')})}}" produces 10:00 AM

"text": "{{TIME(${formatDateTime(date, 'yyyy-MM-ddTHH:mm:ss-00:00')})}}" produces 05:00 AM

It appears that formatDateTime simply substitute the text from the example.

"text": "{{TIME(${formatDateTime(date, 'yyyy-MM-ddTHH:mm:ssZ')})}}" should produces 05:00 AM since Z is same as UCT + 0

**Comment by dclaux** [comment]
Indeed it seems that `formatDateTime` might have a bug where it doesn't handle Z properly.

**Comment by jwoo-msft** [comment]
> Indeed it seems that `formatDateTime` might have a bug where it doesn't handle Z properly.

Yes, I think it's AEL's bug as other offsets work.

**Comment by shalinijoshi19** [comment]
@jwoo-msft this issue can be closed now right?
